carbon-design-system/carbon-addons-iot-react
# 2839 time picker spinner improved keyboard input and added error state

Closes #2839 
Improved keyboard input and added a validation state to the onChangeCallback for the TimePickerSpinner.

It turned our that completely preventing all illegal input without crippling the normal UX (text selection, cut, paste, tab focus etc) would be near impossible. So I made the TimePickerSpinner play well with the invalid prop and added the calculated validation result to the onChange callback so that it could be lifted to the parent component. Hence this is an optional built-in validation that still allow for external validation, or none, to be used.

**Change List (commits, features, bugs, etc)**
TimePickerSpinner
- added validation and adjusted scss
- kept a small part of the initial effort in preventing invalid input, i.e. to only allow numbers and one colon
- fixed old bx-prefix in scss
- fixed story knobs
- added new story
- added default i18n prop merge (turned out it was not needed for my changes but I left it in there)
- added unit tests

DateTimePicker/DateTimePicker2
- fix to disable apply button when there are invalid values in the TimePickerSpinner
- fixed issue in DateTimePicker where the timepickerspinner values from relative spilled over to absolute and the other way around.
- added new unit test and e2e tests
- added restore faketimer to tests 

As a note I also created the following new issues:
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2899
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2893
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2877

**Acceptance Test (how to verify the PR)**
- Verify that new [TimeSpinnerPicker](https://deploy-preview-XXXXX--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-timepickerspinner--with-state-for-inputvalidation) story shows a validation error when an incorrect time is added. It should work for both 12 and 24 hours 

- Verify that it is not possible to, as an end user, enter an invalid time in the TimePickerSpinners we have in the DateTimePicker and DateTimePicker2 without getting the Apply button disabled. Also verify that the the UI with the validation error present looks ok.

**Regression Test (how to make sure this PR doesn't break old functionality)**
- Verify that the old TimeSpinnerPicker stories work as expected
- Verify that the DateTimePicker and DateTimePicker2 stories work as expected



**Things to look for during review**
